servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/AbstractTimeoutHttpFilter.java

```diff
@@ -151,6 +158,62 @@ private Executor contextExecutor(final HttpRequestMetaData requestMetaData,
                 context.executor() : context.ioExecutor();
     }
 
+    // This is responsible for ensuring that we don't abandon the response resources due to the use of Single.timeout.
+    // It does so by retaining a reference to the StreamingHttpResponse in case we receive a cancellation. It will
+    // then attempt to drain the resource. We're protected from multiple subscribes because the TimeoutSingle will
+    // short circuit also on upstream cancellation, although that is not obviously correct behavior.
+    private static final class CleanupSubscriber implements SingleSource.Subscriber<StreamingHttpResponse> {
+
+        private static final AtomicReferenceFieldUpdater<CleanupSubscriber, Object> stateUpdater =
+                AtomicReferenceFieldUpdater.newUpdater(CleanupSubscriber.class, Object.class, "state");
+
+        private static final String COMPLETE = "complete";
+
+        private final SingleSource.Subscriber<? super StreamingHttpResponse> delegate;
+        @Nullable
+        private volatile Object state;
+
+        CleanupSubscriber(final SingleSource.Subscriber<? super StreamingHttpResponse> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void onSubscribe(Cancellable cancellable) {
+            delegate.onSubscribe(() -> {
+                try {
+                    Object current = stateUpdater.getAndSet(this, COMPLETE);
+                    if (current instanceof StreamingHttpResponse) {
+                        clean((StreamingHttpResponse) current);
```

Comment:
if we've propagated the `StreamingHttpResponse` via `delegate.onSuccess()` isn't it possible this will generate a duplicate subscribe? Can you add comments to clarify why this is OK? I see some comments in the test but not clear this covers all scenarios (someone may call cancel() after Subscriber.onSuccess() despite [RS#2.3](https://github.com/reactive-streams/reactive-streams-jvm?tab=readme-ov-file#2.3) and this maybe racy).

Reply:
The TL;DR is that if we get a cancel it passed through the timeout operator and 'won', therefore the message will never be propagated. I added clarifying comments and pointed to the test that ensures the behavior.